### PR TITLE
MA-870 Add/Remove Route fee via Updatecart api endpoint

### PR DESCRIPTION
**Pull request by dumega**
# Description
Add/Remove Route fee via Updatecart api endpoint

Fixes: https://boltpay.atlassian.net/browse/MA-870

#changelog Add/Remove Route fee via Updatecart api endpoint

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

File: ThirdPartyModules/Route/Route.php [hunk 1]
```diff
@@ -41,76 +78,184 @@ class Route
...
+     *
+     * @see \Bolt\Boltpay\Model\Api\UpdateCart::execute
+     * 
+     * @param bool $flag
+     * @param array $removeItem
+     * @param \Magento\Checkout\Model\Session $checkoutSession
+     *
+     * @return bool if return true, then move to next item, if return false, just process the normal execution.
+     */
+    public function filterRemoveItemBeforeUpdateCart($result, $removeItem, $checkoutSession)
```
Comment (danac-gs) [inline ThirdPartyModules/Route/Route.php:222]: `filterRemoveItemBeforeUpdateCart` and `filterAddItemBeforeUpdateCart` can be joined into one method with additional `setInsured` flag parameter. Anyway, it's a matter of preference, not a real requirement.

**Review by danac-gs — CHANGES_REQUESTED** [review]

File: ThirdPartyModules/Route/Route.php [hunk 2]
```diff
@@ -41,76 +78,184 @@ class Route
...
      */
-    public function filterTransactionBeforeOrderCreateValidation($transaction)
+    public function filterCartBeforeLegacyShippingAndTax($cart)
     {
-        $transaction->order->cart->items = array_filter(
-            $transaction->order->cart->items,
+        $parentQuoteId = $cart['order_reference'];
+        $cart['items'] = array_filter(
+            $cart['items'],
             function ($item) {
```
[the remark this answers is not in the document]
Comment (dumega) [reply ThirdPartyModules/Route/Route.php]: Fixed

File: ThirdPartyModules/Route/Route.php [hunk 3]
```diff
@@ -41,76 +78,184 @@ class Route
...
+     * Filter the transaction before Create Order functionality is executed
+     * Remove a dummy product representing the Route Fee additional total, also update cache if Route fee is enabled
+     *
+     * @see \Bolt\Boltpay\Model\Api\CreateOrder::createOrder
+     *
+     * @param array $cart the transaction object from Bolt
+     *
+     * @return array either changed or unchanged $transaction object
+     */
+    public function filterCartBeforeCreateOrder($transaction)
```
[the remark this answers is not in the document]
Comment (dumega) [reply ThirdPartyModules/Route/Route.php:172]: I think we still need to consider other cases that may use other properties of transaction.

File: ThirdPartyModules/Route/Route.php (hunk 1 above)
Comment (dumega) [reply ThirdPartyModules/Route/Route.php:206]: Fixed

**Review by danac-gs — APPROVED** [review]

**Pull request merged**